### ISSUE TITLE
Mode collapsing

### ISSUE THREAD

**Issue by carpedm20** [carpedm20/BEGAN-tensorflow#3]
Couldn't find a reason why, but the model always shows mode collapsing even after learning rate decay one or two times unlike the paper's description. I guess there could be an error on loss but first of all, I don't know how BEGAN tackle this without push-away factor of EBGAN (not sure learning rate decay is enough?).

**Comment by fwiffo** [comment]
May want to try lowering the initial learning rate; it might depend on the dataset. Also, when you're normalizing the images, it doesn't appear that you're centering them on zero. That may also help (subtract 0.5 when normalizing).

Your generated images after 100k steps look quite a bit different than ours; ours tend to be blurry at that point, but yours show a lot of artifacts. So I'm suspecting there is some other bug somewhere.

**Comment by wohnjayne** [comment]
I suspect that the non-linearities in encoder and decoder are not as in the paper
(in both encoder and decoder)
1) slim.conv2d applies a tf.nn.relu() by default, so it should probably read
`x = slim.conv2d(x, hidden_num, 3, 1, data_format=data_format, activation_fn=tf.nn.elu)`
rather than
`x = tf.nn.elu(slim.conv2d(x, hidden_num, 3, 1, data_format=data_format))`
(in both encoder and decoder)
2) in the discriminator the paper says "At the boundary between the encoder and the decoder, the
cube of processed data is mapped via fully connected layers, not followed by any non-linearities", so it should probably be
`z = x = slim.fully_connected(x, z_num, activation_fn=None)`
and
`x = slim.fully_connected(x, np.prod([8, 8, hidden_num]), activation_fn=None)`
3) When "normalizing" the images to center around 0 (by subtracting/adding 0.5), the last conv layer in the decoder shouldnt have an activation function that can actually output negative values, so a `activation_fn=None` might work.

I'm running the above configuration right now, will let you know if it worked.

**Comment by carpedm20** [comment]
@wohnjayne Wow. I really appreciate your help to find this. I'm also really getting stressed.. I once made the exact same mistake before and I still don't understand why they put the default activation function to all the layers. Why..

**Comment by carpedm20** [comment]
@wohnjayne I just ran the experiments below are result only after 4k step. This is completely different from the previous results and I'm pretty sure mode collapsing and all of the bad performance of generator will be fixed. Seriously, many thanks!!

![image](https://cloud.githubusercontent.com/assets/3346407/24960696/7a318e78-1fd1-11e7-91a1-6f60506b1131.png)

**Comment by fwiffo** [comment]
Yes, all of @wohnjayne's observations are correct. The model isn't stable with any hard clipping, is not stable with relu (#1), there are no non-linearities at the autoencoder's hidden (#2), and there is no nonlinearity at the output layer (#3). We tried nonlinearities at the output layer but the model was less stable and had more artifacts. We only clip outputs for producing images, not for calculating the losses.

**Comment by stefdoerr** [comment]
Wow these are some awesome results given that it's still only trained on the celebA dataset. Thanks for the fix!
